File: examples/typescript/vllm_provider_test.ts
#!/usr/bin/env node

/**
 * vLLM Provider Test Example
 *
 * This example demonstrates how to use vLLM models through the Circuit Breaker server's OpenAI API.
 *
 * Prerequisites:
 * - Circuit Breaker server must be running (cargo run --bin server)
 * - vLLM must be configured and available through the server
 *
 * Usage:
 * ```bash
 * # Start the Circuit Breaker server (in another terminal)
 * cargo run --bin server
 *
 * # Run the example
 * npx tsx examples/typescript/vllm_provider_test.ts
 * ```
 */

import fetch from 'node-fetch';
import { Readable } from 'stream';
import { fileURLToPath } from 'url';
import { dirname } from 'path';

const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);

interface ChatMessage {
  role: string;
  content: string;
  name?: string;
}

interface ChatRequest {
  model: string;
  messages: ChatMessage[];
  temperature?: number;
  max_tokens?: number;
  stream?: boolean;
}

interface ChatChoice {
  message?: {
    content?: string;
  };
  delta?: {
    content?: string;
  };
}

interface ChatResponse {
  choices?: ChatChoice[];
  usage?: {
    prompt_tokens?: number;
    completion_tokens?: number;
    total_tokens?: number;
  };
}

interface ModelInfo {
  id: string;
  object?: string;
}

interface ModelsResponse {
  data: ModelInfo[];
}

interface EmbeddingRequest {
  model: string;
  input: string;
  encoding_format?: string;
}

interface EmbeddingData {
  embedding?: number[];
}

interface EmbeddingResponse {
  data?: EmbeddingData[];
}

async function main(): Promise<void> {
  console.log('🚀 Circuit Breaker - vLLM Provider Test');
  console.log('=======================================');

  // Configuration
  const baseUrl = process.env.CIRCUIT_BREAKER_URL || 'http://localhost:3000';
  const openaiEndpoint = `${baseUrl}/v1`;
  const apiKey = process.env.CIRCUIT_BREAKER_API_KEY || '';

  console.log(`🔍 Testing vLLM through Circuit Breaker at: ${openaiEndpoint}`);

  // Test server health
  console.log('\n🏥 Testing server health...');
  try {
    await testHealth(baseUrl);
    console.log('✅ Server is healthy');
  } catch (error) {
    console.error(`❌ Server health check failed: ${error}`);
    return;
  }

  // List available models
  console.log('\n📋 Fetching available models...');
  try {
    const models = await listModels(openaiEndpoint, apiKey);
    console.log('✅ Available models:');
    for (const model of models) {
      console.log(`  - ${model.id}`);
    }
  } catch (error) {
    console.error(`❌ Failed to list models: ${error}`);
    console.log('   This might mean vLLM is not properly configured.');
  }

  // Test chat completion with fallback models
  const model = process.env.VLLM_MODEL || 'microsoft/DialoGPT-medium';
  const fallbackModels = [
    model,
    'microsoft/DialoGPT-medium',
    'codellama/CodeLlama-7b-Instruct-hf',
    'meta-llama/Llama-2-7b-chat-hf',
  ];

  let chatSuccess = false;
  for (const testModel of fallbackModels) {
    console.log(`\n💬 Testing chat completion with model: ${testModel}`);

    const chatRequest: ChatRequest = {
      model: testModel,
      messages: [
        {
          role: 'system',
          content: 'You are a helpful assistant. Be explicit in your responses.'
        },
        {
          role: 'user',
          content: 'Write me an elevator pitch for GitLab'
        }
      ],
      temperature: 0.7,
      max_tokens: 2048,
      stream: false
    };

    try {
      const response = await chatCompletion(openaiEndpoint, apiKey, chatRequest);
      console.log(`✅ Chat completion successful with model: ${testModel}`);
      chatSuccess = true;

      if (response.usage) {
        const { prompt_tokens, completion_tokens, total_tokens } = response.usage;
        if (prompt_tokens && completion_tokens && total_tokens) {
          console.log('📊 Token usage:');
          console.log(`  Prompt tokens: ${prompt_tokens}`);
          console.log(`  Completion tokens: ${completion_tokens}`);
          console.log(`  Total tokens: ${total_tokens}`);
        }
      }

      if (response.choices?.[0]?.message?.content) {
        console.log('\n🤖 Assistant response:');
        console.log(`  ${response.choices[0].message.content}`);
      }
      break; // Success, no need to try other models
    } catch (error) {
      console.error(`❌ Chat completion failed with model '${testModel}': ${error}`);
      if (testModel === fallbackModels[fallbackModels.length - 1]) {
        console.error('   All fallback models failed. Common issues:');
        console.error('   1. No models available in vLLM server');
        console.error('   2. vLLM server not running or misconfigured');
        console.error('   3. Circuit Breaker routing configuration issue');
        console.error('   4. Try setting VLLM_MODEL environment variable to a specific model');
      }
    }
  }

  // Test embeddings (if supported)
  console.log('\n🔮 Testing embeddings...');
  const embeddingModels = [
    process.env.VLLM_EMBEDDING_MODEL || 'sentence-transformers/all-MiniLM-L6-v2',
    'sentence-transformers/all-MiniLM-L6-v2',
    'sentence-transformers/all-mpnet-base-v2',
  ];

  let embeddingSuccess = false;
  for (const embeddingModel of embeddingModels) {
    console.log(`  Trying embedding model: ${embeddingModel}`);

    const embeddingRequest: EmbeddingRequest = {
      model: embeddingModel,
      input: 'This is a test sentence for embeddings.',
      encoding_format: 'float'
    };

    try {
      const response = await embeddings(openaiEndpoint, apiKey, embeddingRequest);
      console.log(`✅ Embeddings successful with model: ${embeddingModel}`);
      embeddingSuccess = true;

      if (response.data?.[0]?.embedding) {
        const embedding = response.data[0].embedding;
        console.log('📊 Embedding details:');
        console.log(`  Model: ${embeddingModel}`);
        console.log(`  Embedding dimension: ${embedding.length}`);

        // Show first 5 values
        const preview = embedding.slice(0, 5);
        console.log(`  First 5 values: [${preview.join(', ')}]`);
      }
      break; // Success, no need to try other models
    } catch (error) {
      console.error(`❌ Embeddings failed with model '${embeddingModel}': ${error}`);
      if (embeddingModel === embeddingModels[embeddingModels.length - 1]) {
        console.error('   All embedding models failed. This might mean:');
        console.error('   1. No embedding models are available in vLLM');
        console.error("   2. vLLM server doesn't support embeddings endpoint");
      }
    }
  }

  // Test streaming (optional)
  if (process.env.TEST_STREAMING === 'true' && chatSuccess) {
    console.log('\n🌊 Testing streaming chat completion...');

    // Use the first successful model from the fallback list
    const streamingModel = fallbackModels[0];

    const streamingRequest: ChatRequest = {
      model: streamingModel,
      messages: [
        {
          role: 'user',
          content: 'Count from 1 to 5, one number per line.'
        }
      ],
      temperature: 0.3,
      max_tokens: 50,
      stream: true
    };

    try {
      await streamingChat(openaiEndpoint, apiKey, streamingRequest);
      console.log('✅ Streaming completed successfully');
    } catch (error) {
      console.error(`❌ Streaming failed: ${error}`);
    }
  } else if (process.env.TEST_STREAMING === 'true') {
    console.log('\n🌊 Skipping streaming test (no successful chat model found)');
  }

  console.log('\n🎉 vLLM provider test completed!');

  // Summary
  console.log('\n📋 Test Summary:');
  console.log(`  Chat completion: ${chatSuccess ? '✅ PASSED' : '❌ FAILED'}`);
  console.log(`  Embeddings: ${embeddingSuccess ? '✅ PASSED' : '❌ FAILED'}`);

  console.log('\n💡 Tips:');
  console.log('  - Set VLLM_MODEL to test specific chat models (default: microsoft/DialoGPT-medium)');
  console.log('  - Set VLLM_EMBEDDING_MODEL to test specific embedding models');
  console.log('  - Set TEST_STREAMING=true to test streaming responses');
  console.log('  - Set CIRCUIT_BREAKER_URL to test remote instances');
  console.log('  - Set CIRCUIT_BREAKER_API_KEY if authentication is required');
  console.log('  - Available fallback models: microsoft/DialoGPT-medium, codellama/CodeLlama-7b-Instruct-hf');
  console.log('  - Make sure vLLM server is running with at least one model loaded');
}

async function testHealth(baseUrl: string): Promise<void> {
  const healthUrl = `${baseUrl}/health`;
  const response = await fetch(healthUrl);

  if (!response.ok) {
    throw new Error(`Health check failed with status: ${response.status}`);
  }
}

async function listModels(baseUrl: string, apiKey: string): Promise<ModelInfo[]> {
  const url = `${baseUrl}/models`;

  const headers: Record<string, string> = {};
  if (apiKey) {
    headers['Authorization'] = `Bearer ${apiKey}`;
  }

  const response = await fetch(url, { headers });

  if (!response.ok) {
    throw new Error(`Models request failed with status: ${response.status}`);
  }

  const json = await response.json() as ModelsResponse;
  return json.data || [];
}

async function chatCompletion(
  baseUrl: string,
  apiKey: string,
  requestBody: ChatRequest
): Promise<ChatResponse> {
  const url = `${baseUrl}/chat/completions`;

  const headers: Record<string, string> = {
    'Content-Type': 'application/json',
  };

  if (apiKey) {
    headers['Authorization'] = `Bearer ${apiKey}`;
  }

  const response = await fetch(url, {
    method: 'POST',
    headers,
    body: JSON.stringify(requestBody),
  });

  if (!response.ok) {
    const errorText = await response.text();
    throw new Error(`Chat completion failed: ${errorText}`);
  }

  return await response.json() as ChatResponse;
}

async function embeddings(
  baseUrl: string,
  apiKey: string,
  requestBody: EmbeddingRequest
): Promise<EmbeddingResponse> {
  const url = `${baseUrl}/embeddings`;

  const headers: Record<string, string> = {
    'Content-Type': 'application/json',
  };

  if (apiKey) {
    headers['Authorization'] = `Bearer ${apiKey}`;
  }

  const response = await fetch(url, {
    method: 'POST',
    headers,
    body: JSON.stringify(requestBody),
  });

  if (!response.ok) {
    const errorText = await response.text();
    throw new Error(`Embeddings failed: ${errorText}`);
  }

  return await response.json() as EmbeddingResponse;
}

async function streamingChat(
  baseUrl: string,
  apiKey: string,
  requestBody: ChatRequest
): Promise<void> {
  const url = `${baseUrl}/chat/completions`;

  const headers: Record<string, string> = {
    'Content-Type': 'application/json',
  };

  if (apiKey) {
    headers['Authorization'] = `Bearer ${apiKey}`;
  }

  const response = await fetch(url, {
    method: 'POST',
    headers,
    body: JSON.stringify(requestBody),
  });

  if (!response.ok) {
    const errorText = await response.text();
    throw new Error(`Streaming chat failed: ${errorText}`);
  }

  const stream = response.body as Readable;
  let buffer = '';

  return new Promise((resolve, reject) => {
    stream.on('data', (chunk: Buffer) => {
      const chunkStr = chunk.toString();
      buffer += chunkStr;

      // Process complete lines from buffer
      let lineEnd;
      while ((lineEnd = buffer.indexOf('\n')) !== -1) {
        const line = buffer.slice(0, lineEnd).trim();
        buffer = buffer.slice(lineEnd + 1);

        if (line.startsWith('data: ')) {
          const data = line.slice(6); // Remove "data: " prefix
          
          if (data === '[DONE]') {
            resolve();
            return;
          }

          try {
            const chunk = JSON.parse(data) as ChatResponse;
            const content = chunk.choices?.[0]?.delta?.content;
            
            if (content) {
              process.stdout.write(content);
            }
          } catch (parseError) {
            // Ignore JSON parse errors for malformed chunks
          }
        }
      }
    });

    stream.on('error', (error) => {
      reject(new Error(`Stream error: ${error.message}`));
    });

    stream.on('end', () => {
      console.log(); // New line after streaming
      resolve();
    });
  });
}

// Run the main function
if (import.meta.url === `file://${process.argv[1]}`) {
  main().catch(console.error);
}

File: examples/rust/vllm_provider_test.rs
//! vLLM Provider Test Example
//!
//! This example demonstrates how to use vLLM models through the Circuit Breaker server's OpenAI API.
//!
//! Prerequisites:
//! - Circuit Breaker server must be running (cargo run --bin server)
//! - vLLM must be configured and available through the server
//!
//! Usage:
//! ```bash
//! # Start the Circuit Breaker server (in another terminal)
//! cargo run --bin server
//!
//! # Run the example
//! cargo run --example vllm_provider_test
//! ```

use reqwest::Client;
use serde_json::{json, Value};
use std::env;

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    // Initialize logging
    tracing_subscriber::fmt::init();

    println!("🚀 Circuit Breaker - vLLM Provider Test");
    println!("=======================================");

    // Configuration
    let base_url =
        env::var("CIRCUIT_BREAKER_URL").unwrap_or_else(|_| "http://localhost:3000".to_string());
    let openai_endpoint = format!("{}/v1", base_url);

    // API key (if required by server configuration)
    let api_key = env::var("CIRCUIT_BREAKER_API_KEY").unwrap_or_default();

    println!(
        "🔍 Testing vLLM through Circuit Breaker at: {}",
        openai_endpoint
    );

    let client = Client::new();

    // Test server health
    println!("\n🏥 Testing server health...");
    match test_health(&client, &base_url).await {
        Ok(()) => println!("✅ Server is healthy"),
        Err(e) => {
            eprintln!("❌ Server health check failed: {}", e);
            return Ok(());
        }
    }

    // List available models
    println!("\n📋 Fetching available models...");
    match list_models(&client, &openai_endpoint, &api_key).await {
        Ok(models) => {
            println!("✅ Available models:");
            for model in models {
                if let Some(id) = model.get("id").and_then(|v| v.as_str()) {
                    println!("  - {}", id);
                }
            }
        }
        Err(e) => {
            eprintln!("❌ Failed to list models: {}", e);
            println!("   This might mean vLLM is not properly configured.");
        }
    }

    // Test chat completion with fallback models
    let model = env::var("VLLM_MODEL").unwrap_or_else(|_| "microsoft/DialoGPT-medium".to_string());

    // Test chat completion with fallback models
    let fallback_models = vec![
        model.clone(),
        "microsoft/DialoGPT-medium".to_string(),
        "codellama/CodeLlama-7b-Instruct-hf".to_string(),
        "meta-llama/Llama-2-7b-chat-hf".to_string(),
    ];

    let mut chat_success = false;
    for test_model in &fallback_models {
        println!("\n💬 Testing chat completion with model: {}", test_model);

        let chat_request = json!({
            "model": test_model,
            "messages": [
                {
                    "role": "system",
                    "content": "You are a helpful assistant. Be explicit in your responses."
                },
                {
                    "role": "user",
                    "content": "Write me an elevator pitch for GitLab"
                }
            ],
            "temperature": 0.7,
            "max_tokens": 2048,
            "stream": false
        });

        // Check if this is a streaming request
        let is_streaming = chat_request.get("stream").and_then(|v| v.as_bool()).unwrap_or(false);
        
        if is_streaming {
            match streaming_chat(&client, &openai_endpoint, &api_key, chat_request.clone()).await {
                Ok(()) => {
                    println!("✅ Streaming chat completion successful with model: {}", test_model);
                    chat_success = true;
                    break;
                }
                Err(e) => {
                    eprintln!(
                        "❌ Streaming chat completion failed with model '{}': {}",
                        test_model, e
                    );
                    if test_model == fallback_models.last().unwrap() {
                        eprintln!("   All fallback models failed for streaming.");
                    }
                }
            }
        }

        match chat_completion(&client, &openai_endpoint, &api_key, chat_request).await {
            Ok(response) => {
                println!("✅ Chat completion successful with model: {}", test_model);
                chat_success = true;

                if let Some(usage) = response.get("usage") {
                    if let (Some(prompt_tokens), Some(completion_tokens), Some(total_tokens)) = (
                        usage.get("prompt_tokens").and_then(|v| v.as_u64()),
                        usage.get("completion_tokens").and_then(|v| v.as_u64()),
                        usage.get("total_tokens").and_then(|v| v.as_u64()),
                    ) {
                        println!("📊 Token usage:");
                        println!("  Prompt tokens: {}", prompt_tokens);
                        println!("  Completion tokens: {}", completion_tokens);
                        println!("  Total tokens: {}", total_tokens);
                    }
                }

                if let Some(choices) = response.get("choices").and_then(|v| v.as_array()) {
                    if let Some(choice) = choices.first() {
                        if let Some(message) = choice.get("message") {
                            if let Some(content) = message.get("content").and_then(|v| v.as_str()) {
                                println!("\n🤖 Assistant response:");
                                println!("  {}", content);
                            }
                        }
                    }
                }
                break; // Success, no need to try other models
            }
            Err(e) => {
                eprintln!(
                    "❌ Chat completion failed with model '{}': {}",
                    test_model, e
                );
                if test_model == fallback_models.last().unwrap() {
                    eprintln!("   All fallback models failed. Common issues:");
                    eprintln!("   1. No models available in vLLM server");
                    eprintln!("   2. vLLM server not running or misconfigured");
                    eprintln!("   3. Circuit Breaker routing configuration issue");
                    eprintln!(
                        "   4. Try setting VLLM_MODEL environment variable to a specific model"
                    );
                }
            }
        }
    }

    // Test embeddings (if supported)
    println!("\n🔮 Testing embeddings...");
    let embedding_models = vec![
        env::var("VLLM_EMBEDDING_MODEL")
            .unwrap_or_else(|_| "sentence-transformers/all-MiniLM-L6-v2".to_string()),
        "sentence-transformers/all-MiniLM-L6-v2".to_string(),
        "sentence-transformers/all-mpnet-base-v2".to_string(),
    ];

    let mut embedding_success = false;
    for embedding_model in &embedding_models {
        println!("  Trying embedding model: {}", embedding_model);

        let embedding_request = json!({
            "model": embedding_model,
            "input": "This is a test sentence for embeddings.",
            "encoding_format": "float"
        });

        match embeddings(&client, &openai_endpoint, &api_key, embedding_request).await {
            Ok(response) => {
                println!("✅ Embeddings successful with model: {}", embedding_model);
                embedding_success = true;

                if let Some(data) = response.get("data").and_then(|v| v.as_array()) {
                    if let Some(embedding_obj) = data.first() {
                        if let Some(embedding) =
                            embedding_obj.get("embedding").and_then(|v| v.as_array())
                        {
                            println!("📊 Embedding details:");
                            println!("  Model: {}", embedding_model);
                            println!("  Embedding dimension: {}", embedding.len());

                            // Show first 5 values
                            let preview: Vec<f64> = embedding
                                .iter()
                                .take(5)
                                .filter_map(|v| v.as_f64())
                                .collect();
                            println!("  First 5 values: {:?}", preview);
                        }
                    }
                }
                break; // Success, no need to try other models
            }
            Err(e) => {
                eprintln!(
                    "❌ Embeddings failed with model '{}': {}",
                    embedding_model, e
                );
                if embedding_model == embedding_models.last().unwrap() {
                    eprintln!("   All embedding models failed. This might mean:");
                    eprintln!("   1. No embedding models are available in vLLM");
                    eprintln!("   2. vLLM server doesn't support embeddings endpoint");
                }
            }
        }
    }

    // Test streaming (optional)
    if env::var("TEST_STREAMING").unwrap_or_default() == "true" && chat_success {
        println!("\n🌊 Testing streaming chat completion...");

        // Use the first successful model from the fallback list
        let streaming_model = fallback_models
            .iter()
            .find(|m| **m != model || chat_success)
            .unwrap_or(&fallback_models[0]);

        let streaming_request = json!({
            "model": streaming_model,
            "messages": [
                {
                    "role": "user",
                    "content": "Count from 1 to 5, one number per line."
                }
            ],
            "temperature": 0.3,
            "max_tokens": 50,
            "stream": true
        });

        match streaming_chat(&client, &openai_endpoint, &api_key, streaming_request).await {
            Ok(()) => println!("✅ Streaming completed successfully"),
            Err(e) => eprintln!("❌ Streaming failed: {}", e),
        }
    } else if env::var("TEST_STREAMING").unwrap_or_default() == "true" {
        println!("\n🌊 Skipping streaming test (no successful chat model found)");
    }

    println!("\n🎉 vLLM provider test completed!");

    // Summary
    println!("\n📋 Test Summary:");
    println!(
        "  Chat completion: {}",
        if chat_success {
            "✅ PASSED"
        } else {
            "❌ FAILED"
        }
    );
    println!(
        "  Embeddings: {}",
        if embedding_success {
            "✅ PASSED"
        } else {
            "❌ FAILED"
        }
    );

    println!("\n💡 Tips:");
    println!(
        "  - Set VLLM_MODEL to test specific chat models (default: microsoft/DialoGPT-medium)"
    );
    println!("  - Set VLLM_EMBEDDING_MODEL to test specific embedding models");
    println!("  - Set TEST_STREAMING=true to test streaming responses");
    println!("  - Set CIRCUIT_BREAKER_URL to test remote instances");
    println!("  - Set CIRCUIT_BREAKER_API_KEY if authentication is required");
    println!("  - Available fallback models: microsoft/DialoGPT-medium, codellama/CodeLlama-7b-Instruct-hf");
    println!("  - Make sure vLLM server is running with at least one model loaded");

    Ok(())
}

async fn test_health(client: &Client, base_url: &str) -> Result<(), Box<dyn std::error::Error>> {
    let health_url = format!("{}/health", base_url);
    let response = client.get(&health_url).send().await?;

    if response.status().is_success() {
        Ok(())
    } else {
        Err(format!("Health check failed with status: {}", response.status()).into())
    }
}

async fn list_models(
    client: &Client,
    base_url: &str,
    api_key: &str,
) -> Result<Vec<Value>, Box<dyn std::error::Error>> {
    let url = format!("{}/models", base_url);

    let mut request = client.get(&url);

    if !api_key.is_empty() {
        request = request.header("Authorization", format!("Bearer {}", api_key));
    }

    let response = request.send().await?;

    if !response.status().is_success() {
        return Err(format!("Models request failed with status: {}", response.status()).into());
    }

    let json: Value = response.json().await?;

    Ok(json
        .get("data")
        .and_then(|v| v.as_array())
        .unwrap_or(&vec![])
        .clone())
}

async fn chat_completion(
    client: &Client,
    base_url: &str,
    api_key: &str,
    request_body: Value,
) -> Result<Value, Box<dyn std::error::Error>> {
    let url = format!("{}/chat/completions", base_url);

    let mut request = client
        .post(&url)
        .header("Content-Type", "application/json")
        .json(&request_body);

    if !api_key.is_empty() {
        request = request.header("Authorization", format!("Bearer {}", api_key));
    }

    let response = request.send().await?;

    if !response.status().is_success() {
        let error_text = response.text().await?;
        return Err(format!("Chat completion failed: {}", error_text).into());
    }

    let json: Value = response.json().await?;
    Ok(json)
}

async fn embeddings(
    client: &Client,
    base_url: &str,
    api_key: &str,
    request_body: Value,
) -> Result<Value, Box<dyn std::error::Error>> {
    let url = format!("{}/embeddings", base_url);

    let mut request = client
        .post(&url)
        .header("Content-Type", "application/json")
        .json(&request_body);

    if !api_key.is_empty() {
        request = request.header("Authorization", format!("Bearer {}", api_key));
    }

    let response = request.send().await?;

    if !response.status().is_success() {
        let error_text = response.text().await?;
        return Err(format!("Embeddings failed: {}", error_text).into());
    }

    let json: Value = response.json().await?;
    Ok(json)
}

async fn streaming_chat(
    client: &Client,
    base_url: &str,
    api_key: &str,
    request_body: Value,
) -> Result<(), Box<dyn std::error::Error>> {
    let url = format!("{}/chat/completions", base_url);

    let mut request = client
        .post(&url)
        .header("Content-Type", "application/json")
        .json(&request_body);

    if !api_key.is_empty() {
        request = request.header("Authorization", format!("Bearer {}", api_key));
    }

    let response = request.send().await?;

    if !response.status().is_success() {
        let error_text = response.text().await?;
        return Err(format!("Streaming chat failed: {}", error_text).into());
    }

    // For streaming, we get the raw text and parse SSE format
    let text = response.text().await?;
    
    println!("\n🤖 Streaming response:");

    // Parse SSE format: "data: {json}\n\n"
    for line in text.lines() {
        if line.starts_with("data: ") {
            let json_str = &line[6..]; // Remove "data: " prefix
            if json_str == "[DONE]" {
                break;
            }

            if let Ok(chunk) = serde_json::from_str::<Value>(json_str) {
                if let Some(choices) = chunk.get("choices").and_then(|v| v.as_array()) {
                    if let Some(choice) = choices.first() {
                        if let Some(delta) = choice.get("delta") {
                            if let Some(content) = delta.get("content").and_then(|v| v.as_str()) {
                                print!("{}", content);
                                std::io::Write::flush(&mut std::io::stdout()).unwrap();
                            }
                        }
                    }
                }
            }
        }
    }

    println!(); // New line after streaming
    Ok(())
}
